src/dragonball/src/dbs_pci/src/root_device.rs

```diff
@@ -0,0 +1,411 @@
+        self.state
+            .read()
+            .expect("poisoned lock for PCI root device")
+            .buses
+            .get(&(id as u32))
+            .cloned()
+    }
+
+    /// Get PCI root device resources
+    #[cfg(target_arch = "aarch64")]
+    pub fn get_device_resources(&self) -> DeviceResources {
+        self.resources.clone()
+    }
+}
+
+impl DeviceIo for PciRootDevice {
+    /// At present, only arm will go to this process, because the x86 pci root device does
+    /// not allocate mmio resources. This process does not involve the arm proprietary
+    /// interface, so it does not use the aarch64 macro to wrap.
+    fn read(&self, base: IoAddress, offset: IoAddress, data: &mut [u8]) {
+        let offset = offset.raw_value();
+        let len = data.len();
+
+        // Only allow naturally aligned Dword, Word and Byte access.
+        if (len == 4 || len == 2 || len == 1) && offset as usize & (len - 1) == 0 {
+            // Do not expect poisoned lock here.
+            let state = self.state.read().unwrap();
+            let (b, d, f, o) = parse_mmio_address(offset);
+            if let Some(bus) = state.buses.get(&b) {
+                return bus.read_config(d, f, o | ((offset as u32) & 0x3), data);
+            }
+        }
+
+        debug!(
+            "Invalid PCI configuration mmio read ({:x}, {})",
+            base.raw_value(),
+            len
+        );
+
+        fill_config_data(data);
+    }
+
+    /// Same as the read interface, currently only this interface is used by the arm.
+    /// For specific reasons, please refer to the comment of the read interface.
+    fn write(&self, base: IoAddress, offset: IoAddress, data: &[u8]) {
+        let offset = offset.raw_value();
+        let len = data.len();
+
+        if (len == 4 || len == 2 || len == 1) && offset as usize & (len - 1) == 0 {
+            // Safe to unwrap because no legal to generate poisoned RwLock.
+            let state = self.state.read().unwrap();
+            let (b, d, f, o) = parse_mmio_address(offset);
+            if let Some(bus) = state.buses.get(&b) {
+                return bus.write_config(d, f, o | (offset as u32 & 0x3), data);
+            }
+        }
+
+        debug!(
+            "Invalid PCI configuration mmio write ({:x}, {})",
+            base.raw_value(),
+            len
+        );
+    }
+
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn pio_read(&self, base: PioAddress, offset: PioAddress, data: &mut [u8]) {
+        let offset = offset.raw_value();
+        let len = data.len();
+
+        // Only allow naturally aligned Dword, Word and Byte access.
+        if (len == 4 || len == 2 || len == 1) && offset as usize & (len - 1) == 0 {
+            match offset {
+                // Configuration address register
+                0..=3 => {
+                    // Do not expect poisoned lock here.
+                    let io_addr = self.state.read().unwrap().io_addr >> (offset << 3);
```

Comment:
Seems like there is a hard code again. Could you leave some comments or set a const?

Reply:
Document added.